keps/sig-network/3698-multi-network/README.md

```diff
@@ -0,0 +1,1114 @@
+|--------------------|----------------------------------------------------------------------------------------------------------------------------------------|
+| ParamsNotReady     | The ParamsReady condition is not present or has “false” value. This error can only happen when the “parametersRefs” field has a value. |
+| PodNetworkNotReady | The referenced PodNetwork object Ready condition has “false” value.                                                                    |
+
+The conditions life-cycle will be handled by the PodNetwork controller described below.
+
+##### InUse indicator
+The PodNetworkAttachment object can be referenced by at least one Pod. When this
+is the case, the PodNetworkAttachment cannot be deleted. This will be maintained
+by the PodNetwork Controller and enforced via a finalizer.\
+When identifying Pods using PodNetworkAttachment for this purpose, the controller
+will filter out Pods that are in Succeeded or Failed state.
+
+##### Mutability
+The PodNetworkAttachment object will be immutable.
+The API server will provide the admission control for that.
+
+##### Lifecycle
+The PodNetworkAttachment will be in following phases:
+1. Created - when the user just created the object and it does not have any
+conditions. 
+2. NotReady - when PodNetworkAttachment’s Ready condition is false. Pods can
+reference such PodNetworkAttachment, but will be in Pending state until the
+PodNetwork becomes Ready. 
+3. Ready - when validation of the PodNetworkAttachment succeeded and Ready
+condition is set to true, here Pod can start attaching to a PodNetworkAttachment. 
+4. InUse - when there is a Pod that references a given PodNetworkAttachment.
+PodNetworkAttachment deletion is blocked by a finalizer when InUse.
+
+##### Validations
+We will introduce following validations for this object:
+* Prevent mutation
+* Ensure podNetworkName is a string
+* Ensure listed parametersRef object fields are strings
+This validation will be performed in the API server.
+
+#### Resources relations
+The relation of the new objects to Pod and between each other is described in
+this diagram:
+<p align="center">
+  <img src="obj-reletion.png?raw=true" alt="introduced objects relationship"/>
+</p>
+
+The arrows define which object references what object. Additionally:
+* PodNetworkAttachment must reference exactly 1 PodNetwork
+* PodNetwork can be referenced by multiple PodNetworkAttachments
+* Pod can reference multiple PodNetworks or PodNetworkAttachments
+* Pod has to reference at least 1 PodNetwork or PodNetworkAttachment (for
+backward compatibility, when nothing specified, “default” PodNetwork is
+auto-populated, see details below)
+* For specific PodNetwork a Pod can either reference that PodNetwork or
+PodNetworkAttachment (referencing that PodNetwork), but not both at the same time
+* PodNetwork can be referenced by multiple Pods
+* PodNetworkAttachment can be referenced by multiple Pods
```

Comment:
In multi-tenancy situations, Pod dimension configuration is usually different.
So that the PodNetworkAttachment data may be 1:1 with the number of Pods.
This will cause a lot of overhead.

如果PodNetworkAttachment 是单个 Pod 维度配置，则没有必要被多个 Pod 引用。

若网络定义不同，则通过多个 PodNetwork 来规划网络平面。

Reply:
This is to provide flexibility to the implementations. One can support 1:1 the other 1:n.

Reply:
We can not regress in scalability and performance, @l1b0k question and comment is spot on, we need to understand the impact of this overhead, we can not sacrifice performance by flexibility without having a real understand on how much penalty will cause this flexibility , and what is the benefit of this flexibility ... open a door for multiple implementations to behave different is not a good answer if this means users will have different behaviors on different places for the same cluster and APIs